fe/fe-core/src/main/java/org/apache/doris/plugin/dialect/HttpDialectUtils.java

```diff
@@ -76,33 +155,199 @@ public static String convertSql(String targetURL, String originStmt, String dial
+                            return null;
                         }
                         return result.data;
                     } else {
-                        LOG.warn("failed to convert sql, response: {}", result);
-                        return originStmt;
+                        LOG.warn("Failed to convert sql, response: {}, URL: {}", result, targetURL);
+                        return null;
                     }
                 }
             } else {
-                LOG.warn("failed to convert sql, response code: {}", responseCode);
-                return originStmt;
+                throw new Exception("HTTP response code: " + responseCode);
             }
-        } catch (Exception e) {
-            LOG.warn("failed to convert sql", e);
-            return originStmt;
         } finally {
             if (connection != null) {
                 connection.disconnect();
             }
         }
     }
 
+    /**
+     * URL Manager - Responsible for URL parsing, caching, blacklist management, and smart selection
+     */
+    private static class UrlManager {
+        private final List<String> parsedUrls;
+        private final ConcurrentHashMap<String, BlacklistEntry> blacklist;
+
+        public UrlManager(String urls) {
+            this.parsedUrls = parseUrls(urls);
+            this.blacklist = new ConcurrentHashMap<>();
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Created UrlManager with URLs: {}, parsed: {}", urls, parsedUrls);
+            }
+        }
+
+        /**
+         * Parse comma separated URL strings
+         */
+        private List<String> parseUrls(String urls) {
+            List<String> result = Lists.newArrayList();
+            if (urls != null && !urls.trim().isEmpty()) {
+                String[] urlArray = urls.split(",");
+                for (String url : urlArray) {
+                    String trimmedUrl = url.trim();
+                    if (!trimmedUrl.isEmpty()) {
+                        result.add(trimmedUrl);
+                    }
+                }
+            }
+            return result;
+        }
+
+        /**
+         * Mark URL as healthy (remove from blacklist)
+         */
+        public void markUrlAsHealthy(String url) {
+            if (blacklist.remove(url) != null) {
+                LOG.info("Removed URL from blacklist due to successful request: {}", url);
+            }
+        }
+
+        /**
+         * Add URL to blacklist
+         */
+        public void markUrlAsBlacklisted(String url) {
+            long currentTime = System.currentTimeMillis();
+            long recoverTime = currentTime + BLACKLIST_RECOVERY_TIME_MS;
+            BlacklistEntry existingEntry = blacklist.get(url);
+            if (existingEntry != null) {
+                // If URL is already in blacklist, limit maximum recovery time to avoid infinite extension
+                // Maximum recovery time is 2 times the original recovery time
```

Comment:
No need, that is too complicated. I think just return if the url is already in black list.